src/buildercore/terraform.py

```diff
@@ -381,7 +378,7 @@ def render_gcs(context):
         },
     }
 
-def render_bigquery(context):
+def render_bigquery(context, tf_file):
```

Comment:
I agree this parameter is too thorny, as it introduces a coupling from a previous `render_*()` (which has to create the `data` key) to this one. I'd rather either do a not-too-clever merge:
```
tf_file = {'resources': {}, 'data': {}}
renders = [render_fastly, render_bigquery, ...]
partials = [r(c) for r in renders]
for p in partials:
    if p['resources']:
        tf_file['resources'] = tf_file['resources'].merge(p['resources'])
    if p['data']:
        tf_file['data'] = tf_file['data'].merge(p['data'])
// remove 'data' and 'resources' if empty
```
More long-term, we could even put each partial in its own `.tf` file as Terraform should be able to resolve them all into a single set of resources, data, etc. That's too involved for this PR however.

Reply:
We should look to cfngen.py and trop.py for how we're handling this pattern already. 

https://github.com/elifesciences/builder/blob/master/src/buildercore/cfngen.py#L133-L156
https://github.com/elifesciences/builder/blob/master/src/buildercore/trop.py#L928-L946

wow - trop.py hasn't gotten *big*. we don't have too many individual files at 1k+ loc

Reply:
`cfngen` doesn't match to a good example in my view, as the `context` it creates its our own data structure, rather than an external one we have to conform to; in fact, it is designed with no overlapping keys e.g. `s3`, `ec2`, ... and the rendering functions are all independent.

`trop` instead passes in a `template` object that the `render_*()` function can write to, while returning `None`. The `tf_file` dictionary however runs the risk that a `render_2()` either:

- overwrites the tf_file['data']['http'] of another `render_1()`
- does lots of null checks `if 'http' in tf_file['data']:` to add to or create the subdictionaries

as there are many overlaps of a single rendering function providing some `resource` and some `data` keys.

`trop` solves this problem by providing a [`Template` class](https://github.com/cloudtools/troposphere/blob/master/troposphere/__init__.py#L552) that can receive `add_resource` and `add_output` calls and be dumped to a dictionary when fully populated. This is a good solution for this case.

Reply:
you know - I almost apologised in my last comment but I take that almost-apology back ;)

objects don't horrify me, they're just almost always the wrong answer to any problem. 

the examples I gave were not meant to be exact fits, but examples of building up a data struct (or object) by passing through it through a simple list of functions. I'll take a look at this chunk of code today.

Reply:
> as there are many overlaps of a single rendering function providing some resource and some data keys.

yes, I was just looking at the code and your `render_fastly` might have it's `resource`s nuked by `render_gcs` that comes next. I mean right now it's probably not a problem as a fastly project doesn't have any buckets (or vice versa). A nice testable version would be the fastly, gcs, bigquery stuff all generating the `data` and `resources` struct that they need and then merging the result afterwards.

Reply:
The long-term version should not need to merge them as they can be stored in different `.tf` files, I'm looking for a safe way to integrate this to avoid divergence in the `buildercore.terraform` module.

Reply:
> A nice testable version would be the fastly, gcs, bigquery stuff all generating the data and resources

I have no horse in this game, so if you implement this procedural solution with some logic in this module, I'll approve it as it's much better than the coupling on `tf_file`

Reply:
I didn't know about indivudlal .tf files, definitely a better solution in the end than more-code. For now though I'll just get something safe and workable happening